### PR TITLE
Add support for custom outcomes and use OkHttp3 for all network calls…

### PR DESCRIPTION
**Pull request by bryansills**
… on all devices

- Use OkHttp3 for all network calls on all devices. The Http stack for Volley [changes](https://github.com/google/volley/blob/master/src/main/java/com/android/volley/toolbox/Volley.java#L54) depending on the devices API version. Using OkHttp3 makes sure that all of the network requests are made using the same code. Also, it is often safer, faster, and more reliable than what Volley uses by default. Also also, you can use Android Studio's new [network profiler](https://developer.android.com/studio/profile/network-profiler.html) with OkHttp3.

Maybe TODO:
- Do a database migration for the new outcome values. For example, the app currently stores the `voicemail` outcome value as `"vm"`, but the new custom outcome value is `"voicemail"`. Work should probably be done so that all values in the database are `"voicemail"`.
- Use Enums for outcomes. They would probably be easier to maintain for associated properties.

### CODE REVIEW

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+import android.widget.Button;
+import android.widget.LinearLayout;
+
+import org.a5calls.android.a5calls.R;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public class OutcomeView extends LinearLayout {
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java]: Why do so much layout work in Java instead of XML? I'd much prefer XML layouts, I find them easier to read and edit than Java (plus Android Studio has a good preview feature).

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            if (orientation == Configuration.ORIENTATION_PORTRAIT) {
+                setOrientation(VERTICAL);
+            } else {
+                setOrientation(HORIZONTAL);
+            }
+
+            buttons = new ArrayList<>();
+            for (final String outcome : outcomes) {
+                Button button = new AppCompatButton(context);
+                button.setMinHeight(dpsToPixels(context, 48));
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java]: Please use R.dimen.accessibility_min_size in dimens.xml rather than hard-coding values.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        if (outcomes != null && outcomes.length > 0) {
+            int orientation = context.getResources().getConfiguration().orientation;
+            if (orientation == Configuration.ORIENTATION_PORTRAIT) {
+                setOrientation(VERTICAL);
+            } else {
+                setOrientation(HORIZONTAL);
+            }
+
+            buttons = new ArrayList<>();
+            for (final String outcome : outcomes) {
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java]: Add a TODO to use a ListView instead of doing this in a for loop. I can't ask you to change it now because I've got code like this elsewhere in the app for the sake of doing stuff quickly, but it's not the best.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    private List<Button> buttons;
+
+    public OutcomeView(Context context, String[] outcomes, final Callback callback) {
+        super(context);
+
+        setLayoutParams(new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
+        setOrientation(VERTICAL);
+
+        if (outcomes != null && outcomes.length > 0) {
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java]: Remove "skip" from outcomes. It doesn't make sense in mobile and @nickoneill has ok'ed this.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/controller/RepCallActivity.java [hunk 5]
```diff
@@ -69,6 +69,7 @@
     private Tracker mTracker = null;
 
     @BindView(R.id.scroll_view) ScrollView scrollView;
+    @BindView(R.id.rep_call_root) LinearLayout linearLayout;
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/controller/RepCallActivity.java]: Give this a better name -- linearLayout is too generic.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/controller/StatsActivity.java [hunk 6]
```diff
@@ -90,9 +91,13 @@ private void initializeUI(DatabaseHelper db) {
         callCountHeader.setText(getStringForCount(
                 mCallCount, R.string.your_call_count_one, R.string.your_call_count));
 
-        List<Long> contacts = db.getCallTimestampsForType(RepCallActivity.CONTACTED);
-        List<Long> voicemails = db.getCallTimestampsForType(RepCallActivity.VOICEMAIL);
-        List<Long> unavailables = db.getCallTimestampsForType(RepCallActivity.UNAVAILABLE);
+        List<Long> contacts = db.getCallTimestampsForType(OutcomeView.CONTACTED);
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/controller/StatsActivity.java]: I think we need to discuss this further with @nickoneill; I've started on Slack. Stats shouldn't have all this special-case work, instead we should have an enum as part of the outcomes JSON that maps each custom outcome to our existing categories. Then we won't log what button the user clicked, instead we will keep logging the type of outcome (vm/skip/contact). Otherwise logs will get diluted with things like "supports" or "opposes" or "agrees", etc, which isn't useful for a user's stats.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/net/OkHttp3Stack.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+import org.apache.http.HttpResponse;
+import org.apache.http.ProtocolVersion;
+import org.apache.http.StatusLine;
+import org.apache.http.entity.BasicHttpEntity;
+import org.apache.http.message.BasicHeader;
+import org.apache.http.message.BasicHttpResponse;
+import org.apache.http.message.BasicStatusLine;
+
+/**
+ * Taken from: https://gist.github.com/NightlyNexus/bdacc5c6951fd9a82b01695e6f6bc600
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/net/OkHttp3Stack.java]: Cool. Is there an open-source license on that gist that we can add to our licenses information?

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java (hunk 1 above)
Comment (bryansills) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java]: It is my understanding that there could be an indeterminate number of outcomes. I'm not sure how to accomplish that with XML layouts other than having an XML file with a bunch of buttons that I programmatically show/hide. I thought this solution would be a little more elegant.

File: 5calls/app/build.gradle [hunk 8]
```diff
@@ -32,6 +35,7 @@ dependencies {
     compile 'com.android.support:design:25.2.0'
     testCompile 'junit:junit:4.12'
     compile 'com.android.volley:volley:1.0.0'
+    compile 'com.squareup.okhttp3:okhttp:3.8.1'
```
Comment (bryansills) [inline 5calls/app/build.gradle]: Done.

[the remark this answers is not in the document]
Comment (bryansills) [reply 5calls/app/build.gradle]: See: https://github.com/5calls/android/pull/89/files#diff-69d7ccf990174d7623a55dece091005cR52

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java (hunk 1 above)
Comment (bryansills) [reply 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java]: Using a RecyclerView now. See:

https://github.com/5calls/android/pull/89/files#diff-4b1b18290a53a1a020c32ca84f6efc7dR1
https://github.com/5calls/android/pull/89/files?diff=split#diff-619663bd5226fe79ebf27b7e14359d94R197
https://github.com/5calls/android/pull/89/files?diff=split#diff-6d9976551a58cdfcf44f3b6a3c33c12aR1

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java (hunk 4 above)
Comment (bryansills) [reply 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java]: Done. See: 
https://github.com/5calls/android/pull/89/files?diff=split#diff-d78ff5861ba70301cb399e336d08cdcbR143
https://github.com/5calls/android/pull/89/files?diff=split#diff-d78ff5861ba70301cb399e336d08cdcbR246

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java (hunk 3 above)
Comment (bryansills) [reply 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java]: Done. See earlier comment about RecyclerView.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java (hunk 2 above)
Comment (bryansills) [reply 5calls/app/src/main/java/org/a5calls/android/a5calls/view/OutcomeView.java]: Fixed in new RecyclerView item layout. See:

https://github.com/5calls/android/pull/89/files?diff=split#diff-6d9976551a58cdfcf44f3b6a3c33c12aR6

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/controller/RepCallActivity.java (hunk 5 above)
Comment (bryansills) [reply 5calls/app/src/main/java/org/a5calls/android/a5calls/controller/RepCallActivity.java]: Removed.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/controller/StatsActivity.java (hunk 6 above)
Comment (bryansills) [reply 5calls/app/src/main/java/org/a5calls/android/a5calls/controller/StatsActivity.java]: Done. See:

https://github.com/5calls/android/pull/89/files?diff=split#diff-a3bc9f38b50e18c62b12e978cb7ad48fR81

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/net/OkHttp3Stack.java (hunk 7 above)
Comment (bryansills) [reply 5calls/app/src/main/java/org/a5calls/android/a5calls/net/OkHttp3Stack.java]: Done. See:

https://github.com/5calls/android/pull/89/files?diff=split#diff-69d7ccf990174d7623a55dece091005cR58
https://github.com/5calls/android/pull/89/files?diff=split#diff-d2c7e22841894f0bd2a901ab4e0dce39R4

**Comment by bryansills** [comment]
All requested changes made.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/model/DatabaseHelper.java [hunk 9]
```diff
@@ -76,10 +73,26 @@ public void onCreate(SQLiteDatabase db) {
 
     @Override
     public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
-        if (newVersion >= 2) {
+        if (oldVersion < 2) {
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/model/DatabaseHelper.java]: newVersion is supplied & can be used to make sure we don't upgrade too far. I like to make these step-wise, making sure not to go past the supplied newVersion:

if (oldVersion == 1 && oldVersion < newVersion) {
 // do upgrade from 1 to 2
 oldVersion = 2;
}
if (oldVersion == 2 && oldVersion < newVersion) {
  // do upgrade from 2 to 3
  oldVersion = 3;
}

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/model/Outcome.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
+package org.a5calls.android.a5calls.model;
+
+import android.content.Context;
+import android.os.Parcel;
+import android.os.Parcelable;
+
+import org.a5calls.android.a5calls.R;
+
+public class Outcome implements Parcelable {
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/model/Outcome.java]: Please add short class comment.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/net/FiveCallsApi.java [hunk 11]
```diff
@@ -231,4 +242,22 @@ private void onRequestError(VolleyError error) {
             Log.d("Error", error.getMessage());
         }
     }
+
+    private void filterSkipOutcomes(List<Issue> issues) {
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/net/FiveCallsApi.java]: This is pretty unrelated to the API. Please refactor into a static method in the Outcome class instead.

File: 5calls/app/src/main/assets/licenses.html [hunk 12]
```diff
@@ -48,6 +48,18 @@
...
+        <br/>
+        <p>Copyright 2017</p>
+        <p>Square Inc.</p>
+    </li>
+    <li>
+        <b>OkHttpStack.java</b>
+        <br/>
+        <p>Copyright 2016</p>
+        <p>Eric Cochran</p>
+    </li>
```
Comment (dektar) [inline 5calls/app/src/main/assets/licenses.html]: Please also add GSON under the apache license: https://github.com/google/gson/blob/master/LICENSE

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/AppSingleton.java [hunk 13]
```diff
@@ -35,8 +45,19 @@ public DatabaseHelper getDatabaseHelper() {
 
     public FiveCallsApi getJsonController() {
         if (mFiveCallsApi == null) {
-            mFiveCallsApi = new FiveCallsApi(mContext);
+            mFiveCallsApi = new FiveCallsApi(getRequestQueue(mContext), getGson());
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/AppSingleton.java]: Because nothing else in the app uses the request queue or Outcome GSON I don't think these functions should be in AppSingleton. Instead, let the FiveCallsApi constructor make the RequestQueue and GSON, abstracted away from the rest of the app.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/adapter/OutcomeAdapter.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public void setEnabled(boolean enabled) {
+        this.enabled = enabled;
+        notifyDataSetChanged();
+    }
+
+    @Override
+    public OutcomeAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_outcome, parent, false));
```
Comment (dektar) [inline 5calls/app/src/main/java/org/a5calls/android/a5calls/adapter/OutcomeAdapter.java]: please wrap lines at 100 chars here and below.

File: 5calls/app/src/main/res/layout/item_outcome.xml [hunk 15]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Button xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@+id/outcome_button"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:minHeight="@dimen/accessibility_min_size"
```
Comment (dektar) [inline 5calls/app/src/main/res/layout/item_outcome.xml]: Could you make the buttons the same look as before? I like the square corners and taller height: add these two lines:

android:background="@color/button_color"
android:layout_marginTop="8dp"

File: 5calls/app/src/main/assets/licenses.html (hunk 12 above)
Comment (bryansills) [reply 5calls/app/src/main/assets/licenses.html]: Already there: https://github.com/bryansills/5calls-android/blob/5df680a0852868928f4e6a9d6a09d67014ed599b/5calls/app/src/main/assets/licenses.html#L28

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/model/DatabaseHelper.java (hunk 9 above)
Comment (bryansills) [reply 5calls/app/src/main/java/org/a5calls/android/a5calls/model/DatabaseHelper.java]: Done.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/net/FiveCallsApi.java (hunk 11 above)
Comment (bryansills) [reply 5calls/app/src/main/java/org/a5calls/android/a5calls/net/FiveCallsApi.java]: Done.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/AppSingleton.java (hunk 13 above)
Comment (bryansills) [reply 5calls/app/src/main/java/org/a5calls/android/a5calls/AppSingleton.java]: We would benefit from using the dependency injection pattern, but since we don't really do any unit testing yet, I'll change it back.

File: 5calls/app/src/main/java/org/a5calls/android/a5calls/adapter/OutcomeAdapter.java (hunk 14 above)
Comment (bryansills) [reply 5calls/app/src/main/java/org/a5calls/android/a5calls/adapter/OutcomeAdapter.java]: Done.

File: 5calls/app/src/main/res/layout/item_outcome.xml (hunk 15 above)
Comment (bryansills) [reply 5calls/app/src/main/res/layout/item_outcome.xml]: Changed the background back to what it was before. Can't use `marginTop` for RecyclerViews. I am using an `ItemDecoration` in order to have the spacing.

**Comment by bryansills** [comment]
All requested changes made.
